File: samples/java/src/app/DoublyLinkedLists.java
package app;

// UI settings
// Expression: linkedList
//   Type: DoublyLinkedLists.DoublyLinkedList
//     Layout: Linked List
//     Nodes: head
//     Save: Object Type Attributes
//   Type: DoublyLinkedLists.Node
//     Layout: Linked List
//     Nodes: next,prev (ctrl+click to select mutliple)
//     Properties: value
//     Save: Object Type Attributes

public class DoublyLinkedLists {

    public static void main(String[] args) {
        // custom linkedlist
        DoublyLinkedList<String> linkedList = new DoublyLinkedList<>();
        for(int i=0; i<10; i++) {
            linkedList.append(String.valueOf(i));
        }

        Node<String> node = linkedList.getHead();
        while (node != null) {
            System.out.println(node.getValue());
            if (node.getValue().equals("5")) {
                node.insertNext(new Node<>("50"));
                break;
            }
            node = node.getNext();
        }

        System.out.println("done");
    }

    private static class DoublyLinkedList<T> {

        protected Node<T> head;
        protected Node<T> tail;

        public Node<T> getHead() {
            return head;
        }

        @SuppressWarnings("unused")
        public Node<T> getTail() {
            return tail;
        }

        public void append(T value) {
            if (head == null) {
                head = new Node<>(value);
                tail = head;
            } else {
                tail.insertNext(new Node<>(value));
                tail = tail.next;
            }
        }
    }

    private static class Node<T> {

        protected T value;
        protected Node<T> next;
        protected Node<T> prev;

        public Node(T value) {
            this.value = value;
        }

        public T getValue() {
            return value;
        }

        public Node<T> getNext() {
            return next;
        }

        @SuppressWarnings("unused")
        public Node<T> getPrev() {
            return prev;
        }

        public void insertNext(Node<T> node) {
            Node<T> t = this.next;
            this.next = node;
            node.next = t;
            if (t != null)
                t.prev = node;
            node.prev = this;
        }

        @Override
        public String toString() {
            return value.toString();
        }
    }
}

File: samples/csharp/src/DoublyLinkedLists.cs
using System;

// UI settings
// Expression: linkedList
//   Type: DoublyLinkedLists.DoublyLinkedList<string>
//     Layout: Linked List
//     Nodes: Head
//     Save: Object Type Attributes
//   Type: DoublyLinkedLists.Node<string>
//     Layout: Linked List
//     Nodes: Next,Prev (ctrl+click to select mutliple)
//     Properties: value
//     Save: Object Type Attributes

public class DoublyLinkedLists
{

    public static void RunMain(string[] args)
    {
        new DoublyLinkedLists().Start();
    }

    public void Start()
    {
        // custom linkedlist
        DoublyLinkedList<string> linkedList = new DoublyLinkedList<string>();
        for (int i = 0; i < 7; i++)
        {
            linkedList.Append(i.ToString());
        }

        Node<string> node = linkedList.Head;
        while (node != null)
        {
            Console.WriteLine(node.Value);
            if (node.Value.Equals("5"))
            {
                node.InsertNext(new Node<string>("50"));
                break;
            }
            node = node.Next;
        }

        Console.WriteLine("done");
    }

    class DoublyLinkedList<T>
    {

        public Node<T> Head { get; private set; }
        public Node<T> Tail { get; private set; }

        public void Append(T value)
        {
            if (Head == null)
            {
                Head = new Node<T>(value);
                Tail = Head;
            }
            else
            {
                Tail.InsertNext(new Node<T>(value));
                Tail = Tail.Next;
            }
        }
    }

    private class Node<T>
    {

        public T Value { get; private set; }
        public Node<T> Next { get; private set; }
        public Node<T> Prev { get; private set; }

        public Node(T value)
        {
            this.Value = value;
        }

        public void InsertNext(Node<T> node)
        {
            Node<T> t = this.Next;
            this.Next = node;
            node.Next = t;
            if (t != null)
                t.Prev = node;
            node.Prev = this;
        }

        public override string ToString()
        {
            return Value.ToString();
        }
    }
}